docs/man/memstat.1
.TH "MEMSTAT" "1" "September 28, 2011" "0.53" "libmemcached"
.SH NAME
memstat \- libmemcached Documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.\" Man page generated from reStructeredText.
.
.sp
memstat \- Display the operating status of a single or group of memcached servers
.SH SYNOPSIS
.sp
memstat [options]
.SH DESCRIPTION
.sp
\fBmemstat\fP  dumps the state of memcached(1) servers.
It displays all data to stdout.
.SH OPTIONS
.sp
You can specify servers via the option:
.INDENT 0.0
.TP
.B \-\-servers
.UNINDENT
.sp
or via the environment variable:
.INDENT 0.0
.TP
.B \(aqMEMCACHED_SERVERS, \-\-args\(aq
.UNINDENT
.sp
which can be used to specify the "argument" sent to the stats command (ie slab, size, items, etc..).
.sp
For a full list of operations run the tool with:
.INDENT 0.0
.TP
.B \-\-help
.UNINDENT
.SH HOME
.sp
To find out more information please check:
\fI\%http://libmemcached.org/\fP
.SH SEE ALSO
.sp
\fImemcached(1)\fP \fIlibmemcached(3)\fP
.SH AUTHOR
Brian Aker
.SH COPYRIGHT
2011, Brian Aker DataDifferential, http://datadifferential.com/
.\" Generated by docutils manpage writer.
.\"
.


ChangeLog
2024-09-02 [version 1.14.0]
## 🐣 New Features
* Add `mop upsert` method. #292
* Add multiple storage method.
  * `memcached_mset()`
  * `memcached_madd()`
  * `memcached_mreplace()`
  * `memcached_mprepend()`
  * `memcached_mpppend()`
  * `memcached_mcas()`
## 🐛 Bug Fixes
* Fix buffer size when calling `snprintf()` with leading whitespace and integer value. #304
* Use `__linux__` instead of `TARGET_OS_LINUX` to check if the current OS is Linux. #306
* Use different lock filenames for multiple proxy in a process
## 📝 Documentation
* Fix typo.
## ✅ Testing
* Cancel previous CI in progress when new commits are added. #296

2024-02-22 [version 1.13.3]
## 🔧 Enhancements
* Fix compile warnings on RHEL8 and MacOS.
* Do not use `#pragma once`.
* Select smallest host as owner when ketama hash collided. #237
* Add parameter validation in `arcus_proxy_connect()`.
* Move io reset from where calls `memcached_vdo()` to where fails inside of `memcached_vdo()`. #250
* Use `<latest-release>-unknown` version when git describe fails.
* Add a limit of count about bop mget/smget commands.
* Import `scanner.l` and `parser.yy` from libmemcached-0.53.
* Refactor ketama weighted related implementation. #263 #273 #274 #275 #276 #277
* Fix package name and info to 'arcus-c-client'.
## 🐛 Bug Fixes
* Call `arcus_server_check_for_update()` in `memcached_exist_by_key()`. #241
* Check `mc->connect_timeout` instead of `mc->poll_timeout` in `connect_poll()`. #247
* Remove a duplicate init macro in configure.ac. #267
* Return `hashkit_get_distribution_function()` instead of `hashkit_get_function()` in `memcached_behavior_get_distribution_hash()`.
## 📝 Documentation
* Add unit test FAQ.
* Fix inconsistencies between code and documentation.
## ✅ Testing
* Use Github Actions instead of Travis for CI test.

2022-02-22 [version 1.13.2]
        * [FEATURE] supported the 3 replicas in replication.
        * [FIX] return switchover error code when coll get with delete.
        * [FIX] fixed wrong aggregate pipe return code in case of failure.
        * [FIX] hold the pool lock when update the cache list of master mc.
        * [FIX] fixed the cache list update of member mc in proxy mode.
        * [ENHANCE] enhanced the cachelist update logic with serverinfo.
        * [ENHANCE] Not to close and reopen pool connections
        * [ENHANCE] removed the dual pool repopulation when start at first.
        * [ENHANCE] optimized the slave comparision in memcached_rgroup_update().
        * [ENHANCE] optimized the release processing in memcached_pool.
        * [ENHANCE] ensure more concurrency in memcached pool.
        * [CLEANUP] fixed the compile warnings.
        * [CLEANUP] refactored serveral parts of code.

2021-07-06 [version 1.13.1]
        * [FEATURE] enable zookeeper multi thread mode.
        * [FIX] if MEMCACHED_ERRNO entered in memcached_set_error(),
                return instead of assert.

2021-01-18 [version 1.13.0]
        * [FEATURE] supported mgets command.
        * [FIX] check space_sepatated_keys immediately after memcached_connect().
        * [FIX] check mget command support immediately after memcached_connect().
        * [INTERNAL] return MEMCACHED_CONNECTION_FAILURE in memcached_fetch_result().

2020-12-09 [version 1.12.0]
        * [FEATURE] set the max key size to 4000.
        * [DOC] changed max element bytes: 4KB => 16KB.
                (committed at 20200317 and included in 1.10.4)
        * [DOC] fixed the smget result API and structure names.

2020-12-07 [version 1.11.0]
        * [FEATURE] supported space-separated keys.
        * [FEATURE] supported mget command, efficient for getting multiple items.
        * [FEATURE] requested version operation when connect to server.
        * [FEATURE] increased max key length: 250 => 4K.
        * [ENHANCE] do memcached_quit_server() if poll timeout occurs in io_flush.
        * [ENHANCE] increased cluster max size for multi key op: 200 => 1000.
        * [IMPROVE] reduced string copy when convert eflag/update_filter to string.
        * [FIX] set purge recursion flag to false before return.
        * [FIX] handled the host failures in the multi key operation.
        * [FIX] unset noreply before get operation in noreply_test().
        * [FIX] fixed unallocated memory access in mkey fetch.
        * [OTHERS] Refactored some code for readabiblity.

2020-10-22 [version 1.10.5]
        * [FEATURE] bop incr/decr with initial option.
        * [FIX] registered "NULL" if hostname lookup fails in do_add_client_info().
        * [FIX] added the TYPE_MISMATCH handling in kv response
        * [DOC] changed the description of the max item size
        * [DOC] refactored the document format and some content.

2020-08-04 [version 1.10.4]
        * [FEATURE] Added memcached_detail_error_message() API.
        * [FEATURE] Immediate reconnect according to error responses.
        * [FEATURE] Changed max element bytes. 4KB => 16KB.
        * [FEATURE] Returned E2BIG if CLIENT_ERROR object too large.
        * [FEATURE] Changed MEMCACHED_SERVER_FAILURE_RETRY_TIMEOUT. 2sec => 1sec.
        * [ENHANCE] Enlarged the contents of error messages in detail.
        * [ENHANCE] Stored detail error message in error_messages field.
        * [ENHANCE] Added arcus ZK manager for processing ZK events.
        * [FIX] pipe operation is stopped if switchover is done.
        * [FIX] Fixed version check condition.
        * [FIX] Fixed several bugs in handling response strings.
        * [FIX] Added the key test where needed.
        * [FIX] Do client_info and watcher registration only when needed.
        * [FIX] Prevent master_mc from being freed while using pool.
        * [FIX] Changed to create version using m4 file.
        * [FIX] Fixed simple null pointer dereferencing.
        * [DOC] Added multi eflag filter description.
        * [Test] Added test case for PIPE_ERROR bad error.
        * [OTHERS] Refactored the response module for handling responses.

clients/memstat.cc
/* LibMemcached
 * Copyright (C) 2006-2009 Brian Aker
 * All rights reserved.
 *
 * Use and distribution licensed under the BSD license.  See
 * the COPYING file in the parent directory for full text.
 *
 * Summary:
 *
 * Authors:
 *          Brian Aker
 *          Toru Maesaka
 */
#include <config.h>

#include <cstdio>
#include <cstring>
#include <ctime>
#include <iostream>
#include <fcntl.h>
#include <getopt.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/types.h>

#include <libmemcached/memcached.h>

#include "client_options.h"
#include "utilities.h"

#define PROGRAM_NAME "memstat"
#define PROGRAM_DESCRIPTION "Output the state of a memcached cluster."

/* Prototypes */
static void options_parse(int argc, char *argv[]);
static void run_analyzer(memcached_st *memc, memcached_stat_st *memc_stat);
static void print_analysis_report(memcached_st *memc,
                                  memcached_analysis_st *report);

static int opt_verbose= 0;
static int opt_displayflag= 0;
static int opt_analyze= 0;
static char *opt_servers= NULL;
static char *stat_args= NULL;
static char *analyze_mode= NULL;

static struct option long_options[]=
{
  {(OPTIONSTRING)"args", required_argument, NULL, OPT_STAT_ARGS},
  {(OPTIONSTRING)"version", no_argument, NULL, OPT_VERSION},
  {(OPTIONSTRING)"help", no_argument, NULL, OPT_HELP},
  {(OPTIONSTRING)"quiet", no_argument, NULL, OPT_QUIET},
  {(OPTIONSTRING)"verbose", no_argument, &opt_verbose, OPT_VERBOSE},
  {(OPTIONSTRING)"debug", no_argument, &opt_verbose, OPT_DEBUG},
  {(OPTIONSTRING)"servers", required_argument, NULL, OPT_SERVERS},
  {(OPTIONSTRING)"flag", no_argument, &opt_displayflag, OPT_FLAG},
  {(OPTIONSTRING)"analyze", optional_argument, NULL, OPT_ANALYZE},
  {0, 0, 0, 0},
};


static memcached_return_t stat_printer(memcached_server_instance_st instance,
                                       const char *key, size_t key_length,
                                       const char *value, size_t value_length,
                                       void *context)
{
  static memcached_server_instance_st last= NULL;
  (void)context;

  if (last != instance)
  {
    printf("Server: %s (%u)\n", memcached_server_name(instance),
           (uint32_t)memcached_server_port(instance));
    last= instance;
  }

  printf("\t %.*s: %.*s\n", (int)key_length, key, (int)value_length, value);

  return MEMCACHED_SUCCESS;
}

int main(int argc, char *argv[])
{
  options_parse(argc, argv);
  initialize_sockets();

  if (opt_servers == NULL)
  {
    char *temp;
    if ((temp= getenv("MEMCACHED_SERVERS")))
    {
      opt_servers= strdup(temp);
    }
    else
    {
      std::cerr << "No Servers provided" << std::endl;
      return EXIT_FAILURE;
    }
  }

  memcached_st *memc= memcached_create(NULL);

  memcached_server_st *servers= memcached_servers_parse(opt_servers);
  free(opt_servers);

  memcached_return_t rc= memcached_server_push(memc, servers);
  memcached_server_list_free(servers);

  if (rc != MEMCACHED_SUCCESS && rc != MEMCACHED_SOME_ERRORS)
  {
    printf("Failure to communicate with servers (%s)\n",
           memcached_strerror(memc, rc));
    exit(1);
  }

  if (opt_analyze)
  {
    memcached_stat_st *memc_stat;

    memc_stat= memcached_stat(memc, NULL, &rc);

    if (! memc_stat)
      exit(-1);

    run_analyzer(memc, memc_stat);

    memcached_stat_free(memc, memc_stat);
  }
  else
  {
    rc= memcached_stat_execute(memc, stat_args, stat_printer, NULL);
  }

  memcached_free(memc);

  return rc == MEMCACHED_SUCCESS ? EXIT_SUCCESS: EXIT_FAILURE;
}

static void run_analyzer(memcached_st *memc, memcached_stat_st *memc_stat)
{
  memcached_return_t rc;

  if (analyze_mode == NULL)
  {
    memcached_analysis_st *report;
    report= memcached_analyze(memc, memc_stat, &rc);
    if (rc != MEMCACHED_SUCCESS || report == NULL)
    {
      printf("Failure to analyze servers (%s)\n",
             memcached_strerror(memc, rc));
      exit(1);
    }
    print_analysis_report(memc, report);
    free(report);
  }
  else if (strcmp(analyze_mode, "latency") == 0)
  {
    uint32_t flags, server_count= memcached_server_count(memc);
    uint32_t num_of_tests= 32;
    const char *test_key= "libmemcached_test_key";

    memcached_st **servers;
    servers= static_cast<memcached_st**>(malloc(sizeof(memcached_st*) * server_count));
    if (not servers)
    {
      fprintf(stderr, "Failed to allocate memory\n");
      return;
    }

    for (uint32_t x= 0; x < server_count; x++)
    {
      memcached_server_instance_st instance=
        memcached_server_instance_by_position(memc, x);

      if ((servers[x]= memcached_create(NULL)) == NULL)
      {
        fprintf(stderr, "Failed to memcached_create()\n");
        if (x > 0)
          memcached_free(servers[0]);
        x--;

        for (; x > 0; x--)
          memcached_free(servers[x]);

        free(servers);
        return;
      }
      memcached_server_add(servers[x],
                           memcached_server_name(instance),
                           memcached_server_port(instance));
    }

    printf("Network Latency Test:\n\n");
    struct timeval start_time, end_time;
    uint32_t slowest_server= 0;
    long elapsed_time, slowest_time= 0;

    for (uint32_t x= 0; x < server_count; x++)
    {
      memcached_server_instance_st instance=
        memcached_server_instance_by_position(memc, x);
      gettimeofday(&start_time, NULL);

      for (uint32_t y= 0; y < num_of_tests; y++)
      {
        size_t vlen;
        char *val= memcached_get(servers[x], test_key, strlen(test_key),
                                 &vlen, &flags, &rc);
        if (rc != MEMCACHED_NOTFOUND && rc != MEMCACHED_SUCCESS)
          break;
        free(val);
      }
      gettimeofday(&end_time, NULL);

      elapsed_time= (long) timedif(end_time, start_time);
      elapsed_time /= (long) num_of_tests;

      if (elapsed_time > slowest_time)
      {
        slowest_server= x;
        slowest_time= elapsed_time;
      }

      if (rc != MEMCACHED_NOTFOUND && rc != MEMCACHED_SUCCESS)
      {
        printf("\t %s (%d)  =>  failed to reach the server\n",
               memcached_server_name(instance),
               memcached_server_port(instance));
      }
      else
      {
        printf("\t %s (%d)  =>  %ld.%ld seconds\n",
               memcached_server_name(instance),
               memcached_server_port(instance),
               elapsed_time / 1000, elapsed_time % 1000);
      }
    }

    if (server_count > 1 && slowest_time > 0)
    {
      memcached_server_instance_st slowest=
        memcached_server_instance_by_position(memc, slowest_server);

      printf("---\n");
      printf("Slowest Server: %s (%d) => %ld.%ld seconds\n",
             memcached_server_name(slowest),
             memcached_server_port(slowest),
             slowest_time / 1000, slowest_time % 1000);
    }
    printf("\n");

    for (uint32_t x= 0; x < server_count; x++)
      memcached_free(servers[x]);

    free(servers);
    free(analyze_mode);
  }
  else
  {
    fprintf(stderr, "Invalid Analyzer Option provided\n");
    free(analyze_mode);
  }
}

static void print_analysis_report(memcached_st *memc,
                                  memcached_analysis_st *report)

{
  uint32_t server_count= memcached_server_count(memc);
  memcached_server_instance_st most_consumed_server= memcached_server_instance_by_position(memc, report->most_consumed_server);
  memcached_server_instance_st least_free_server= memcached_server_instance_by_position(memc, report->least_free_server);
  memcached_server_instance_st oldest_server= memcached_server_instance_by_position(memc, report->oldest_server);

  printf("Memcached Cluster Analysis Report\n\n");

  printf("\tNumber of Servers Analyzed         : %u\n", server_count);
  printf("\tAverage Item Size (incl/overhead)  : %u bytes\n",
         report->average_item_size);

  if (server_count == 1)
  {
    printf("\nFor a detailed report, you must supply multiple servers.\n");
    return;
  }

  printf("\n");
  printf("\tNode with most memory consumption  : %s:%u (%llu bytes)\n",
         memcached_server_name(most_consumed_server),
         (uint32_t)memcached_server_port(most_consumed_server),
         (unsigned long long)report->most_used_bytes);
  printf("\tNode with least free space         : %s:%u (%llu bytes remaining)\n",
         memcached_server_name(least_free_server),
         (uint32_t)memcached_server_port(least_free_server),
         (unsigned long long)report->least_remaining_bytes);
  printf("\tNode with longest uptime           : %s:%u (%us)\n",
         memcached_server_name(oldest_server),
         (uint32_t)memcached_server_port(oldest_server),
         report->longest_uptime);
  printf("\tPool-wide Hit Ratio                : %1.f%%\n", report->pool_hit_ratio);
  printf("\n");
}

static void options_parse(int argc, char *argv[])
{
  memcached_programs_help_st help_options[]=
  {
    {0},
  };

  int option_index= 0;

  bool opt_version= false;
  bool opt_help= false;
  while (1)
  {
    int option_rv= getopt_long(argc, argv, "Vhvds:a", long_options, &option_index);

    if (option_rv == -1)
      break;

    switch (option_rv)
    {
    case 0:
      break;

    case OPT_VERBOSE: /* --verbose or -v */
      opt_verbose = OPT_VERBOSE;
      break;

    case OPT_DEBUG: /* --debug or -d */
      opt_verbose = OPT_DEBUG;
      break;

    case OPT_VERSION: /* --version or -V */
      opt_version= true;
      break;

    case OPT_HELP: /* --help or -h */
      opt_help= true;
      break;

    case OPT_SERVERS: /* --servers or -s */
      opt_servers= strdup(optarg);
      break;

    case OPT_STAT_ARGS:
      stat_args= strdup(optarg);
      break;

    case OPT_ANALYZE: /* --analyze or -a */
      opt_analyze= OPT_ANALYZE;
      analyze_mode= (optarg) ? strdup(optarg) : NULL;
      break;

    case OPT_QUIET:
      close_stdio();
      break;

    case '?':
      /* getopt_long already printed an error message. */
      exit(1);
    default:
      abort();
    }
  }

  if (opt_version)
  {
    version_command(PROGRAM_NAME);
    exit(EXIT_SUCCESS);
  }

  if (opt_help)
  {
    help_command(PROGRAM_NAME, PROGRAM_DESCRIPTION, long_options, help_options);
    exit(EXIT_SUCCESS);
  }
}
